upcloud/managed_database.go

```diff
@@ -0,0 +1,693 @@
+		if len(m.Rows[i]) != len(m.Columns) {
+			return nil, fmt.Errorf("unexpected number of columns at row %d (not %d)", i, len(m.Columns))
+		}
+		for j := range m.Rows[i] {
+			row = append(row, &m.Rows[i][j])
+		}
+		rows = append(rows, row)
+	}
+	chart.Data.Rows = rows
+	return json.Marshal(&chart)
+}
+
+// UnmarshalJSON implements json.Unmarshaler for the canonical form of metrics chart
+func (m *ManagedDatabaseMetricsChartInt) UnmarshalJSON(d []byte) error {
+	var chart canonicalManagedDatabaseMetricChartForUnmarshal
+	if err := json.Unmarshal(d, &chart); err != nil {
+		return err
+	}
+	timestamps := make([]time.Time, 0, len(chart.Data.Rows))
+	rows := make([][]int, 0, len(chart.Data.Rows))
+	inRows := chart.Data.Rows
+	for i := range inRows {
+		var row = make([]int, 0, len(inRows)-1)
+		var ts time.Time
+		if len(inRows[i]) != len(chart.Data.Columns) {
+			return fmt.Errorf("unexpected number of columns at row %d (not %d)", i, len(m.Columns))
+		}
+		if err := json.Unmarshal(inRows[i][0], &ts); err != nil {
+			return fmt.Errorf("cannot unmarshal timestamp at row %d: %w", i, err)
+		}
+		timestamps = append(timestamps, ts)
+		for j := 1; j < len(inRows[i]); j++ {
+			var iv int
+			if err := json.Unmarshal(inRows[i][j], &iv); err != nil {
+				return fmt.Errorf("cannot unmarshal value at row %d, column %d: %w", i, j, err)
+			}
+			row = append(row, iv)
+		}
+		rows = append(rows, row)
+	}
+	m.Timestamps = timestamps
+	m.Columns = chart.Data.Columns[1:]
+	m.Title = chart.Hints.Title
+	m.Rows = rows
+	return nil
+}
+
+// ManagedDatabaseMetricsColumn represents a single column of a metrics chart
+type ManagedDatabaseMetricsColumn struct {
+	// Label describes the usage of chart's column
+	Label string `json:"label"`
+	// Type describes the type of values in chart's column
+	Type string `json:"type"`
+}
+
+// ManagedDatabaseNodeState represents a database node that is part of the service instance
+type ManagedDatabaseNodeState struct {
+	// Name field is same as the ManagedDatabase.Name plus a dash plus an index value. The index represents the
+	// generation of a node. Certain modifications require re-provisioning of a node.
+	Name string `json:"name"`
+	// Role represents the role of a node
+	// master  = The node servers read and write requests
+	// standby = The node servers read-only requests and is ready to assume the master role during failure scenario
+	Role  string `json:"role"`
+	State string `json:"state"`
+}
+
+// ManagedDatabaseProperties is a Properties helper type for ManagedDatabase
+type ManagedDatabaseProperties map[string]interface{}
+
+// Get returns a property value by name. The underlying map is initialised if it's nil
+func (m *ManagedDatabaseProperties) Get(name string) interface{} {
+	if m == nil {
+		return nil
+	}
+	if *m == nil {
+		*m = make(ManagedDatabaseProperties)
+	}
+	return (*m)[name]
+}
+
+// GetString returns a string property value. An empty string is returned for non-existing or not-a-string values.
```

Comment:
I'm a bit iffy on the 'return default value  on wrong-type values' approach, seems like it could lead to errors that are very easy to miss. Would returning an error in that case make the other bits too unwieldy?

Reply:
I took inspiration from Vipers's GetString method. Obviously there's more complex semantics there of overlapping set of maps. Here we could just leave this as plain map[string]interface{}. Then again it would make working with slices more difficult for a user (for example the IPFilter property).

What do you think?

Reply:
For Viper, the 'empty on invalid type' probably makes more sense than for us - for us, it kinda seems to me like it would always be more or less problematic (what if the types change in the input, without us knowing/noticing? Everything would still *seem* to work without a hitch but wouldn't)? I see your point about IPFilter being quite difficult to use if we drop these, maybe just make these return errors in case the type cast fails? 
  
Could even add `Must*`-style methods (eg. panic on error) as well, allowing the consumer to choose their required level of confidence and convenience (and hopefully making them think explicitly about the consequences!) in the types they are receiving?

Reply:
Ok, good idea. Let's use errors with getters and consider adding must getters too.

The property system schema is somewhat dynamic but still they should be predictable. I would like to code generate them at some point but currently the situation is too much in flux for that.

Reply:
Sounds good, code generation would get big :+1: :from me so hope we get there at some point!

Reply:
implemented